File: forssea_utilities/src/forssea_utilities/grafcet.py
#!/usr/bin/env python
from abc import abstractmethod
from collections import deque
from enum import Enum
from threading import Thread
from threading import Lock
import rospy as rp


class GrafcetCondition(object):
    """
    Abstract class modeling a condition to be checked before passing a transition
    """
    def __init__(self, cond_id, grafcet):
        """"
        Constructor
        """
        self._id = cond_id
        self._grafcet = grafcet

    @abstractmethod
    def eval(self):
        """
        The abstract method to be implemented by the user : actual condition expression
        """
        pass


class GrafcetTransition(object):
    """
    Class modeling a transition between two steps
    """
    def __init__(self, condition, upstream_step_ids, downstream_step_ids):
        """
        Constructor
        Note : the step ids are initialized once and for all, whereas the step will be added by the grafcet
        """
        self.__condition = condition
        self.__upstream_step_ids = upstream_step_ids
        self.__upstream_steps = []
        self.__downstream_step_ids = downstream_step_ids
        self.__downstream_steps = []

    def set_condition(self, condition):
        self.__condition = condition

    def add_upstream_step(self, step):
        self.__upstream_steps.append(step)

    def get_upstream_step_ids(self):
        return self.__upstream_step_ids

    def add_downstream_step(self, step):
        self.__downstream_steps.append(step)

    def get_downstream_step_ids(self):
        return self.__downstream_step_ids

    def get_downstream_steps(self):
        return self.__downstream_steps

    def is_enabled(self):
        """
        A transition gets enabled once all its upstream steps are active
        """
        enabled = True
        for upstream_step in self.__upstream_steps:
            enabled = upstream_step.is_active()
            if not enabled:
                break

        return enabled

    def can_be_cleared(self):
        """
        A transition can be cleared if the related condition returns True
        """
        if self.__condition is None:
            return False

        return self.__condition.eval()


class ActionMode(Enum):
    RISING_EDGE = 1,
    CONTINUE = 2,
    FALLING_EDGE = 3


class Action(object):
    """
    Class modeling an action
    """
    def __init__(self, grafcet):
        self._initialized = False
        self._grafcet = grafcet

    def is_initialized(self):
        return self.__initialized

    @abstractmethod
    def initialize(self):
        self._initialized = True
        pass

    @abstractmethod
    def execute(self):
        pass


class GrafcetStep(object):
    """
    Class modeling a step
    """
    def __init__(self, step_id, transitions, action, action_mode=ActionMode.RISING_EDGE):
        """
        Constructor
        """
        self.__id = step_id
        self.__active = False
        self.__transitions = transitions
        self.__action = action
        self.__action_mode = action_mode

    def get_id(self):
        return self.__id

    def get_transitions(self):
        return self.__transitions

    def is_active(self):
        return self.__active

    def set_active(self, active):
        """
        The actual step activation (i.e. action execution) depends on the action mode defined
        """
        if self.__active == active and active:
            if self.__action_mode == ActionMode.CONTINUE:
                self._action()
        else:
            if self.__action_mode == ActionMode.RISING_EDGE and active:
                self._action()
            elif self.__action_mode == ActionMode.FALLING_EDGE and not active:
                self._action()
            elif self.__action_mode == ActionMode.CONTINUE and active:
                self._action(True)

            self.__active = active

    def evolve(self):
        """
        Check whether a downstream transition can be cleared and return the downstream steps
        Note : Only one transition can be cleared at a time
        """
        evolved = False
        downstream_steps = []
        for transition in self.__transitions:
            if transition.is_enabled() and transition.can_be_cleared():
                downstream_steps = transition.get_downstream_steps()
                evolved = True
                break

        return evolved, downstream_steps

    def _action(self, first_activation=False):
        """
        Execute the private action
        """
        if self.__action is None:
            return False

        rp.logdebug("GrafcetStep::_action() : executing S" + str(self.get_id()))
        return self.__action.execute()


class Grafcet(Thread):
    """
    Class modeling a grafcet structure and behavior
    """
    def __init__(self, name, frequency):
        self.__name = name
        self.__frequency = frequency  # Hz
        self.__steps = []
        self.__transitions = []

        self.__active_steps = []

        self._timer = None

        # Diagnostic purpose
        self._diag_step_ids = deque([], 5)
        self._ds_lock = Lock()

        self._setup()

    @property
    def frequency(self):
        return self.__frequency

    def get_step(self, step_id):
        step_list = filter(lambda x: x.get_id() == step_id, self.__steps)
        return step_list[0]

    def _add_step(self, step):

        if step is None:
            return

        # Make sure the step is not already existing
        step_id = step.get_id()
        for existing_step in self.__steps:
            if existing_step.get_id() == step_id:
                # Can never happen except when designing the grafcet
                return

        self.__steps.append(step)

        # If the step is active, add it to the corresponding list
        if step.is_active():
            self.__active_steps.append(step)

        # Record the step's transitions
        self.__transitions.extend(step.get_transitions())

    def _setup(self):
        # Call the specific_setup method, implemented by specific grafcets
        self._specific_setup()
        # Bind upstream and downstream steps to transitions
        self.__bind_transition_to_step()

    @abstractmethod
    def _specific_setup(self):
        pass

    def run(self):

        self._timer = rp.Timer(rp.Duration(1.0 / self.__frequency), lambda x: self._evolve())

    def _evolve(self):

        completed_steps = []
        new_active_steps = []

        for active_step in self.__active_steps:
            # Can the current step evolve?
            evolved, downstream_steps = active_step.evolve()
            if evolved:
                completed_steps.append(active_step)

                new_active_steps.extend(downstream_steps)

            else:
                # The step remains an active step
                new_active_steps.append(active_step)

        # Update the step states
        for completed_step in completed_steps:
            completed_step.set_active(False)

        # Update active_steps while removing duplicated
        self.__active_steps = list(set(new_active_steps))

        for active_step in self.__active_steps:
            active_step.set_active(True)

        # Diagnostics
        self._ds_lock.acquire()
        self._diag_step_ids.append(self.__active_steps[0].get_id())
        self._ds_lock.release()

    def __bind_transition_to_step(self):

        for transition in self.__transitions:
            step_ids = transition.get_upstream_step_ids()
            for step_id in step_ids:
                step = self.get_step(step_id)
                if step is not None:
                    transition.add_upstream_step(step)

            step_ids = transition.get_downstream_step_ids()
            for step_id in step_ids:
                step = self.get_step(step_id)
                if step is not None:
                    transition.add_downstream_step(step)

    def get_diag_step_ids(self):
        self._ds_lock.acquire()
        ds_ids = list(self._diag_step_ids)
        self._ds_lock.release()
        return ds_ids

    @abstractmethod
    def shutdown(self):
        pass

File: forssea_utilities/src/grafcet.cpp
//
// Created by david on 14/05/2020.
//

#include <forssea_utilities/grafcet.hpp>
using namespace forssea_utilities;

//
// GrafcetTransition implementation
//
GrafcetTransition::GrafcetTransition(GrafcetCondition* condition, list<int> upstream_step_ids, list<int> downstream_step_ids)
  : condition_(condition)
{
  upstream_step_ids_ = upstream_step_ids;
  downstream_step_ids_ = downstream_step_ids;
}

GrafcetTransition::~GrafcetTransition()
{
  if (condition_ != 0) {
    delete condition_;
    condition_ = 0;
  }

  upstream_steps_.clear();
  downstream_steps_.clear();
}

bool GrafcetTransition::isEnabled()
{
  ///Check if all upstream steps are actives
  ///if one of them is not active, then transition is not enabled
  list<GrafcetStep*>::iterator sIt = upstream_steps_.begin();
  list<GrafcetStep*>::iterator sItEnd = upstream_steps_.end();
  while (sIt != sItEnd) {
    if (!(*sIt)->isActive())
      return false;
    sIt++;
  }
  return true;
}

bool GrafcetTransition::canBeCleared()
{
  if (condition_ == 0)
    return false;

  return condition_->eval();
}

//
// GrafcetStep implementation
//
GrafcetStep::GrafcetStep(const int& id, list<GrafcetTransition*> transitions, Action* action, ActionMode mode)
    : id_(id), active_(false), mode_(mode), action_(action)
{
  list<GrafcetTransition*>::iterator tIt = transitions.begin();
  list<GrafcetTransition*>::iterator tItEnd = transitions.end();
  while (tIt != tItEnd) {
    transitions_.push_back(*tIt);
    tIt++;
  }
}

GrafcetStep::~GrafcetStep()
{
  this->setActive(false);

  ///Delete action (aggregation)
  if (action_ != 0) {
    delete action_;
    action_ = 0;
  }
}

void GrafcetStep::setActive(bool state)
{
  if (active_ == state  && state) {
    if (mode_ == ActionMode::CONTINUE)
      this->action();
  }
  else {
    if (mode_ == ActionMode::RISING_EDGE && state)
      this->action();
    else if (mode_ == ActionMode::FALLING_EDGE && !state)
      this->action();
    else if (mode_ == ActionMode::CONTINUE && state)
      this->action(true);

    active_ = state;
  }
}

bool GrafcetStep::evolve(list<GrafcetStep*>& downstream_steps)
{
  // Check whether a downstream transition can be cleared and return the downstream steps
  // Note : Only one transition can be cleared at a time
  list<GrafcetTransition*>::iterator tIt = transitions_.begin();
  list<GrafcetTransition*>::iterator tItEnd = transitions_.end();

  while (tIt != tItEnd)
  {
    if ((*tIt)->isEnabled()) {
      if ((*tIt)->canBeCleared()) {
        ///Only one transition can be cleared at a time
        (*tIt)->getDownstreamSteps(downstream_steps);
        return true;
      }
    }
    tIt++;
  }

  return false;
}

bool GrafcetStep::action(bool first_activation)
{
  if (action_ == 0)
    return true;

  return action_->execute();
}

//
// Grafcet implementation
//
Grafcet::Grafcet(const string& name, ros::NodeHandle* node_handle, const float& frequency)
 : frequency_(1.0)
{
  node_handle_ = node_handle;
  name_ = name;
  if (frequency > 0.0) frequency_ = frequency;
}

Grafcet::~Grafcet()
{
  ///Clear the active step list
  active_steps_.clear();

  ///Delete all steps
  list<GrafcetStep*>::iterator sIt = steps_.begin();
  list<GrafcetStep*>::iterator sItEnd = steps_.end();
  while (sIt != sItEnd) {
    if (*sIt != 0) {
      delete *sIt;
      *sIt = 0;
    }
    sIt++;
  }
  steps_.clear();

  ///Delete all transitions
  list<GrafcetTransition*>::iterator tIt = transitions_.begin();
  list<GrafcetTransition*>::iterator tItEnd = transitions_.end();
  while (tIt != tItEnd) {
    if (*tIt != 0) {
      delete *tIt;
      *tIt = 0;
    }
    tIt++;
  }
  transitions_.clear();
}

GrafcetStep* Grafcet::getStep(const int& id)
{
  GrafcetStep* step = 0;

  list<GrafcetStep*>::iterator sIt = steps_.begin();
  list<GrafcetStep*>::iterator sItEnd = steps_.end();
  while (sIt != sItEnd) {
    if ((*sIt)->getId() == id) {
      step = *sIt;
      break;
    }

    sIt++;
  }

  return step;

}

void Grafcet::setup()
{
  //Call specificSetup implemented by specific grafcets
  this->specificSetup();

  ///Bind up/downstreamSteps to transitions
  this->bindTransitionToStep();

  return;
}

void Grafcet::run()
{
  if ( node_handle_ != nullptr ) {
    grafcet_scheduler_ = node_handle_->createTimer(ros::Duration(1 / frequency_), &Grafcet::evolve, this, false, true);
  }
}

list<int> Grafcet::getDiagStepIds()
{
  list<int> ds_ids;
  ds_lock_.lock();
  ds_ids = diag_step_ids_;
  ds_lock_.unlock();
  return ds_ids;
}

void Grafcet::addStep(GrafcetStep* step)
{
  list<GrafcetStep*>::iterator sIt = steps_.begin();
  list<GrafcetStep*>::iterator sItEnd = steps_.end();
  while (sIt != sItEnd) {
    if ((*sIt)->getId() == step->getId()) {
      // Can never happen except when designing the grafcet
      return;
    }

    sIt++;
  }

  steps_.push_back(step);

  ///If added step is active, push back it into the activeSteps_ list
  if (step->isActive())
  {
    active_steps_.push_back(step);
  }

  //Add transitions to transitions_ (memory management)
  list<GrafcetTransition*>::iterator tIt = step->getTransitions().begin();
  list<GrafcetTransition*>::iterator tItEnd = step->getTransitions().end();
  while (tIt != tItEnd)
  {
    transitions_.push_back(*tIt);
    tIt++;
  }
}

void Grafcet::evolve(const ros::TimerEvent &event)
{
  list<GrafcetStep*> completed_steps;
  list<GrafcetStep*> new_active_step_list;

  list<GrafcetStep*>::iterator sIt = active_steps_.begin();
  list<GrafcetStep*>::iterator sItEnd = active_steps_.end();
  while (sIt != sItEnd)
  {
    ///Can the current step evolve ?
    list<GrafcetStep*> downstream_steps;
    if ((*sIt)->evolve(downstream_steps))
    {
      completed_steps.push_back(*sIt);

      //Test if step exist
      list<GrafcetStep*>::iterator sItDownstreamSteps = downstream_steps.begin();
      list<GrafcetStep*>::iterator sItEndDownstreamSteps = downstream_steps.end();
      while(sItDownstreamSteps != sItEndDownstreamSteps)
      {
        bool isExist = false;
        list<GrafcetStep*>::iterator sItnewActiveStepList = new_active_step_list.begin();
        list<GrafcetStep*>::iterator sItEndnewActiveStepList = new_active_step_list.end();
        while(sItnewActiveStepList != sItEndnewActiveStepList && !isExist)
        {
          if(( *sItDownstreamSteps )->getId() == ( *sItnewActiveStepList )->getId())
            isExist = true;
          sItnewActiveStepList++;
        }

        //Add step if not exist
        if( !isExist )
          new_active_step_list.insert( new_active_step_list.end(), *sItDownstreamSteps );

        sItDownstreamSteps++;
      }

    }
    else
      new_active_step_list.push_back(*sIt);

    sIt++;
  }

  ///Desactivate completed step
  sIt = completed_steps.begin();
  sItEnd = completed_steps.end();
  while (sIt != sItEnd)
  {
    (*sIt)->setActive(false);
    sIt++;
  }

  ///Update active steps list
  active_steps_ = new_active_step_list;

  ///Activate active steps
  sIt = active_steps_.begin();
  sItEnd = active_steps_.end();
  while (sIt != sItEnd)
  {
    (*sIt)->setActive(true);
    sIt++;
  }

  ds_lock_.lock();
  diag_step_ids_.push_back(active_steps_.front()->getId());
  if ((int)diag_step_ids_.size() > 5) diag_step_ids_.pop_front();
  ds_lock_.unlock();
}

void Grafcet::bindTransitionToStep()
{
  GrafcetStep* step = 0;
  list<int> step_ids;

  list<GrafcetTransition*>::iterator tIt = transitions_.begin();
  list<GrafcetTransition*>::iterator tItEnd = transitions_.end();
  while (tIt != tItEnd)
  {
    step_ids = (*tIt)->getUpstreamStepIds();
    list<int>::iterator iIt = step_ids.begin();
    list<int>::iterator iItEnd = step_ids.end();
    while (iIt != iItEnd)
    {
      step = this->getStep(*iIt);
      if (step != 0)
        (*tIt)->addUpstreamStep(step);

      iIt++;
    }

    step_ids.clear();
    step_ids = (*tIt)->getDownstreamStepIds();
    iIt = step_ids.begin();
    iItEnd = step_ids.end();
    while (iIt != iItEnd)
    {
      step = this->getStep(*iIt);
      if (step != 0)
        (*tIt)->addDownstreamStep(step);

      iIt++;
    }

    step_ids.clear();

    tIt++;
  }
}

vector<int> Grafcet::getActiveStepIds()
{
  vector<int> active_step_ids;
  list<GrafcetStep*>::iterator gsIt = active_steps_.begin();
  list<GrafcetStep*>::iterator gsItEnd = active_steps_.end();
  while (gsIt != gsItEnd)
  {
    active_step_ids.push_back((*gsIt)->getId());
    gsIt++;
  }

  return active_step_ids;
}
